### PR TITLE
Refactor `OmegaConfigLoader`

### PR DESCRIPTION
**Pull request by merelcht**
## Description
The methods in `OmegaConfigLoader` are big and getting more complicated each time we make a change.

## Development notes
I simplified things by:

- Extracting the filesystem and protocol setup into a separate function `_initialise_filesystem_and_protocol`
- Extracting the merging steps of the `base` and `default_run_env` config into a separate function `_merge_configs`
- Added a check to return early if chosen env config is the same as base config to avoid loading the same config twice and not do unnecessary merging.

## Developer Certificate of Origin
We need all contributions to comply with the [Developer Certificate of Origin (DCO)](https://developercertificate.org/). All commits must be signed off by including a `Signed-off-by` line in the commit message. [See our wiki for guidance](https://github.com/kedro-org/kedro/wiki/Guidelines-for-contributing-developers/).

If your PR is blocked due to unsigned commits, then you must follow the instructions under "Rebase the branch" on the GitHub Checks page for your PR. This will retroactively add the sign-off to all unsigned commits and allow the DCO check to pass.

## Checklist

- [ ] Read the [contributing](https://github.com/kedro-org/kedro/blob/main/CONTRIBUTING.md) guidelines
- [ ] Signed off each commit with a [Developer Certificate of Origin (DCO)](https://docs.github.com/en/repositories/managing-your-repositorys-settings-and-features/managing-repository-settings/managing-the-commit-signoff-policy-for-your-repository)
- [ ] Opened this PR as a 'Draft Pull Request' if it is work-in-progress
- [ ] Updated the documentation to reflect the code changes
- [ ] Added a description of this change in the [`RELEASE.md`](https://github.com/kedro-org/kedro/blob/main/RELEASE.md) file
- [ ] Added tests to cover my changes
- [ ] Checked if this change will affect Kedro-Viz, and if so, communicated that with the Viz team

### CODE REVIEW

File: kedro/config/omegaconf_config.py [hunk 1]
```diff
@@ -358,6 +345,43 @@ def load_and_merge_dir_config(
...
+        return fs, protocol
+
+    def _merge_configs(
+        self,
+        config: dict[str, Any],
+        env_config: dict[str, Any],
+        key: str,
+        env_path: str,
+    ) -> Any:
+        merging_strategy = self.merge_strategy.get(key, "destructive")
```
Comment (ElenaKhaustova) [inline kedro/config/omegaconf_config.py:374]: Shall we go even further and move merging strategies outside to the `Enum` and create a mapping of symbolic names to the implementations? Then all the possible values and implementations will be in one place, so it's easier to navigate and compare them, and additionally, we will get rid `if else` logic when applying a merging strategy.

**Review by ElenaKhaustova — COMMENTED** [review]
Thank you, @merelcht, changes looks good to me.

Left one non-blocking suggestion.

File: kedro/config/omegaconf_config.py (hunk 1 above)
Comment (ElenaKhaustova) [reply kedro/config/omegaconf_config.py:374]: Some clarification on the above comment.

I was thinking of something like that:

```
class MergingStrategies(Enum):
    FIRST = "first"
    SECOND = "second"

def first_strat():
    pass
    
def second_strat():
    pass
    
merging_strategies_map = {
    MergingStrategies.FIRST: first_strat,
    MergingStrategies.SECOND: second_strat,
}
```

Usage:
```
first = strategies_map[MergingStrategies.FIRST]()
second = strategies_map[MergingStrategies("second")]()
```

**Review by ElenaKhaustova — APPROVED** [review]
Thank you, @merelcht, all clear! ✨

**Review by ankatiyar — APPROVED** [review]
Looks good! 💯

**Pull request merged**